File: modules/static-face-detection/ios/StaticFaceDetectionModule.swift
import ExpoModulesCore
import MLKitFaceDetection

struct DetectedFace: Record {
  @Field var bounds: BoundsRecord
  @Field var rollAngle: Double
  @Field var yawAngle: Double
  @Field var pitchAngle: Double
  @Field var landmarks: LandmarksRecord?
  @Field var smilingProbability: Double?
  @Field var leftEyeOpenProbability: Double?
  @Field var rightEyeOpenProbability: Double?
}

struct BoundsRecord: Record {
  @Field var x: Double
  @Field var y: Double
  @Field var width: Double
  @Field var height: Double
}

struct LandmarksRecord: Record {
  @Field var LEFT_EYE: PointRecord?
  @Field var RIGHT_EYE: PointRecord?
  @Field var NOSE_BASE: PointRecord?
  @Field var MOUTH_LEFT: PointRecord?
  @Field var MOUTH_RIGHT: PointRecord?
  @Field var MOUTH_BOTTOM: PointRecord?
  @Field var LEFT_CHEEK: PointRecord?
  @Field var RIGHT_CHEEK: PointRecord?
  @Field var LEFT_EAR: PointRecord?
  @Field var RIGHT_EAR: PointRecord?
}

struct PointRecord: Record {
  @Field var x: Double
  @Field var y: Double
}

struct DetectFacesResult: Record {
  @Field var faces: [DetectedFace]
  @Field var imageWidth: Double
  @Field var imageHeight: Double
}

public class StaticFaceDetectionModule: Module {
  public func definition() -> ModuleDefinition {
    Name("StaticFaceDetectionModule")

    AsyncFunction("detectFaces") { (imageUri: String, promise: Promise) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let url = URL(string: imageUri) ?? URL(fileURLWithPath: imageUri)
          let data = try Data(contentsOf: url)
          guard let image = UIImage(data: data) else {
            promise.reject("ERR_LOAD", "Cannot load image: \(imageUri)")
            return
          }

          let visionImage = VisionImage(image: image)
          let options = FaceDetectorOptions()
          options.performanceMode = .fast
          options.landmarkMode = .all
          options.classificationMode = .all
          options.minFaceSize = 0.15

          let detector = FaceDetector.faceDetector(options: options)
          detector.process(visionImage) { faces, error in
            if let error = error {
              promise.reject("ERR_DETECT", error.localizedDescription)
              return
            }

            let faceList = (faces ?? []).map { face -> DetectedFace in
              let bounds = BoundsRecord(
                x: Double(face.frame.origin.x),
                y: Double(face.frame.origin.y),
                width: Double(face.frame.width),
                height: Double(face.frame.height)
              )

              var landmarks = LandmarksRecord()
              if let lm = face.landmarks?[.leftEye] {
                landmarks.LEFT_EYE = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.rightEye] {
                landmarks.RIGHT_EYE = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.noseBase] {
                landmarks.NOSE_BASE = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.mouthLeft] {
                landmarks.MOUTH_LEFT = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.mouthRight] {
                landmarks.MOUTH_RIGHT = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.mouthBottom] {
                landmarks.MOUTH_BOTTOM = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.leftCheek] {
                landmarks.LEFT_CHEEK = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.rightCheek] {
                landmarks.RIGHT_CHEEK = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.leftEar] {
                landmarks.LEFT_EAR = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }
              if let lm = face.landmarks?[.rightEar] {
                landmarks.RIGHT_EAR = PointRecord(x: Double(lm.position.x), y: Double(lm.position.y))
              }

              return DetectedFace(
                bounds: bounds,
                rollAngle: Double(face.headEulerAngleZ),
                yawAngle: Double(face.headEulerAngleY),
                pitchAngle: Double(face.headEulerAngleX),
                landmarks: landmarks,
                smilingProbability: face.smilingProbability.map { Double($0) },
                leftEyeOpenProbability: face.leftEyeOpenProbability.map { Double($0) },
                rightEyeOpenProbability: face.rightEyeOpenProbability.map { Double($0) }
              )
            }

            let result = DetectFacesResult(
              faces: faceList,
              imageWidth: Double(image.size.width),
              imageHeight: Double(image.size.height)
            )
            promise.resolve(result.toDictionary())
          }
        } catch {
          promise.reject("ERR_LOAD", error.localizedDescription)
        }
      }
    }
  }
}

File: modules/static-face-detection/android/src/main/java/expo/modules/staticfacedetection/StaticFaceDetectionModule.kt
package expo.modules.staticfacedetection

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import android.net.Uri
import androidx.exifinterface.media.ExifInterface
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.io.InputStream

class StaticFaceDetectionModule : Module() {
  private val context
    get() = appContext.reactContext ?: throw Exception("No Android context")

  // Most photos store orientation in EXIF instead of baking it into pixels.
  // decodeStream ignores EXIF, so faces in rotated photos would be missed or
  // mispositioned — rotate the bitmap to display orientation before detection.
  private fun applyExifOrientation(uri: Uri, bitmap: Bitmap): Bitmap {
    var exifStream: InputStream? = null
    return try {
      exifStream = context.contentResolver.openInputStream(uri)
      if (exifStream == null) return bitmap
      val exif = ExifInterface(exifStream)
      val matrix = Matrix()
      when (exif.getAttributeInt(
        ExifInterface.TAG_ORIENTATION,
        ExifInterface.ORIENTATION_NORMAL
      )) {
        ExifInterface.ORIENTATION_ROTATE_90 -> matrix.postRotate(90f)
        ExifInterface.ORIENTATION_ROTATE_180 -> matrix.postRotate(180f)
        ExifInterface.ORIENTATION_ROTATE_270 -> matrix.postRotate(270f)
        ExifInterface.ORIENTATION_FLIP_HORIZONTAL -> matrix.postScale(-1f, 1f)
        ExifInterface.ORIENTATION_FLIP_VERTICAL -> matrix.postScale(1f, -1f)
        ExifInterface.ORIENTATION_TRANSPOSE -> {
          matrix.postScale(-1f, 1f); matrix.postRotate(270f)
        }
        ExifInterface.ORIENTATION_TRANSVERSE -> {
          matrix.postScale(-1f, 1f); matrix.postRotate(90f)
        }
        else -> return bitmap
      }
      Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    } catch (e: Exception) {
      bitmap
    } finally {
      try { exifStream?.close() } catch (_: Exception) {}
    }
  }

  override fun definition() = ModuleDefinition {
    Name("StaticFaceDetectionModule")

    AsyncFunction("detectFaces") { imageUri: String, promise: Promise ->
      try {
        val uri = Uri.parse(imageUri)
        val inputStream = context.contentResolver.openInputStream(uri)
        if (inputStream == null) {
          promise.reject("ERR_OPEN", "Cannot open image: $imageUri", null)
          return@AsyncFunction
        }
        val rawBitmap: Bitmap? = BitmapFactory.decodeStream(inputStream)
        inputStream.close()
        if (rawBitmap == null) {
          promise.reject("ERR_DECODE", "Cannot decode image: $imageUri", null)
          return@AsyncFunction
        }
        val bitmap = applyExifOrientation(uri, rawBitmap)

        val image = InputImage.fromBitmap(bitmap, 0)
        val options = FaceDetectorOptions.Builder()
          .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
          .setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
          .setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
          .setMinFaceSize(0.1f)
          .build()

        val detector = FaceDetection.getClient(options)
        detector.process(image)
          .addOnSuccessListener { faces ->
            val faceList = faces.map { face ->
              val bounds = mapOf(
                "x" to face.boundingBox.left.toDouble(),
                "y" to face.boundingBox.top.toDouble(),
                "width" to face.boundingBox.width().toDouble(),
                "height" to face.boundingBox.height().toDouble()
              )
              val landmarks = mutableMapOf<String, Map<String, Double>>()
              val landmarkTypes = listOf(
                FaceLandmark.LEFT_EYE to "LEFT_EYE",
                FaceLandmark.RIGHT_EYE to "RIGHT_EYE",
                FaceLandmark.NOSE_BASE to "NOSE_BASE",
                FaceLandmark.MOUTH_LEFT to "MOUTH_LEFT",
                FaceLandmark.MOUTH_RIGHT to "MOUTH_RIGHT",
                FaceLandmark.MOUTH_BOTTOM to "MOUTH_BOTTOM",
                FaceLandmark.LEFT_CHEEK to "LEFT_CHEEK",
                FaceLandmark.RIGHT_CHEEK to "RIGHT_CHEEK",
                FaceLandmark.LEFT_EAR to "LEFT_EAR",
                FaceLandmark.RIGHT_EAR to "RIGHT_EAR"
              )
              for ((type, name) in landmarkTypes) {
                val lm = face.getLandmark(type)
                if (lm != null) {
                  landmarks[name] = mapOf(
                    "x" to lm.position.x.toDouble(),
                    "y" to lm.position.y.toDouble()
                  )
                }
              }
              mapOf(
                "bounds" to bounds,
                "rollAngle" to face.headEulerAngleZ.toDouble(),
                "yawAngle" to face.headEulerAngleY.toDouble(),
                "pitchAngle" to face.headEulerAngleX.toDouble(),
                "landmarks" to landmarks.ifEmpty { null },
                "smilingProbability" to face.smilingProbability?.toDouble(),
                "leftEyeOpenProbability" to face.leftEyeOpenProbability?.toDouble(),
                "rightEyeOpenProbability" to face.rightEyeOpenProbability?.toDouble()
              )
            }
            val result = mapOf(
              "faces" to faceList,
              "imageWidth" to bitmap.width.toDouble(),
              "imageHeight" to bitmap.height.toDouble()
            )
            bitmap.recycle()
            if (bitmap !== rawBitmap) rawBitmap.recycle()
            promise.resolve(result)
          }
          .addOnFailureListener { e ->
            bitmap.recycle()
            if (bitmap !== rawBitmap) rawBitmap.recycle()
            promise.reject("ERR_DETECT", e.message, e)
          }
      } catch (e: Exception) {
        promise.reject("ERR_EXCEPTION", e.message, e)
      }
    }
  }
}
